IllegalMonitorStateException when using synchronized(object) { object.notifyAll(); }

I'm trying to test a simple lock structure using model checking. I have found that code of the form:

    synchronized (object) {
        object.notifyAll()
    }

results in an `IllegalMonitorStateException` as the current thread is apparently not the owner of the object's monitor. I would not have expected this exception as the object's monitor is clearly held at the time.

I have attached a piece of demonstration code. The exact output from Lincheck 2.12 is:

    Exception in thread "main" org.jetbrains.kotlinx.lincheck.LincheckAssertionError: 
    = The execution failed with an unexpected exception =
    Execution scenario (parallel part):
    | increment() |

    java.lang.IllegalMonitorStateException
        at java.lang.Object.notifyAll(Native Method)
        at DemoLincheckIssue$SimpleLock.unlock(DemoLincheckIssue.java:29)
        at DemoLincheckIssue.increment(DemoLincheckIssue.java:48)
        at org.jetbrains.kotlinx.lincheck.runner.TestThreadExecution48.run(Unknown Source)
        at org.jetbrains.kotlinx.lincheck.runner.FixedActiveThreadsExecutor$testThreadRunnable$1.run(FixedActiveThreadsExecutor.kt:163)
        at java.lang.Thread.run(Thread.java:748)


    = The following interleaving leads to the error =
    Parallel part trace:
    | increment()                                                                                              |
    |   lock.READ: SimpleLock@1 at DemoLincheckIssue.increment(DemoLincheckIssue.java:45)                      |
    |   lock() at DemoLincheckIssue.increment(DemoLincheckIssue.java:45)                                       |
    |   state.READ: 0 at DemoLincheckIssue.increment(DemoLincheckIssue.java:46)                                |
    |   state.WRITE(1) at DemoLincheckIssue.increment(DemoLincheckIssue.java:46)                               |
    |   state.READ: 1 at DemoLincheckIssue.increment(DemoLincheckIssue.java:47)                                |
    |   lock.READ: SimpleLock@1 at DemoLincheckIssue.increment(DemoLincheckIssue.java:48)                      |
    |   unlock(): threw IllegalMonitorStateException at DemoLincheckIssue.increment(DemoLincheckIssue.java:48) |
    |     MONITORENTER at DemoLincheckIssue$SimpleLock.unlock(DemoLincheckIssue.java:27)                       |
    |     owner.WRITE(null) at DemoLincheckIssue$SimpleLock.unlock(DemoLincheckIssue.java:28)                  |
    |     MONITOREXIT at DemoLincheckIssue$SimpleLock.unlock(DemoLincheckIssue.java:27)                        |
        at org.jetbrains.kotlinx.lincheck.LinChecker.check(LinChecker.kt:51)
        at org.jetbrains.kotlinx.lincheck.LinChecker$Companion.check(LinChecker.kt:182)
        at org.jetbrains.kotlinx.lincheck.LinChecker$Companion.check$default(LinChecker.kt:181)
        at org.jetbrains.kotlinx.lincheck.LinChecker.check(LinChecker.kt)
        at DemoLincheckIssue.main(DemoLincheckIssue.java:53)

I seem to get the same output with Lincheck 2.11.

[DemoLincheckIssue.txt](https://github.com/Kotlin/kotlinx-lincheck/files/6098680/DemoLincheckIssue.txt)

--- a/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt
+++ b/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt
@@ -224,8 +224,10 @@ abstract class ManagedStrategy(
         check(sameResultTypes && sameResults) {
             StringBuilder().apply {
                 appendln("Non-determinism found. Probably caused by non-deterministic code (WeakHashMap, Object.hashCode, etc).")
-                appendln("Reporting scenario without execution trace.")
-                appendln(loggedResults.asLincheckFailureWithoutTrace().toString())
+                appendln("== Reporting the first execution without execution trace ==")
+                appendln(failingResult.asLincheckFailureWithoutTrace())
+                appendln("== Reporting the second execution ==")
+                appendln(loggedResults.toLincheckFailure(scenario, Trace(traceCollector!!.trace, testCfg.verboseTrace)).toString())
             }.toString()
         }
         return Trace(traceCollector!!.trace, testCfg.verboseTrace)
@@ -447,7 +449,7 @@ abstract class ManagedStrategy(
      * @return whether lock should be actually acquired
      */
     internal fun beforeLockAcquire(iThread: Int, codeLocation: Int, tracePoint: MonitorEnterTracePoint?, monitor: Any): Boolean {
-        if (inIgnoredSection(iThread)) return true
+        if (!isTestThread(iThread)) return true
         newSwitchPoint(iThread, codeLocation, tracePoint)
         // Try to acquire the monitor
         if (!monitorTracker.acquireMonitor(iThread, monitor)) {
@@ -455,7 +457,7 @@ abstract class ManagedStrategy(
             // Switch to another thread and wait for a moment when the monitor can be acquired
             switchCurrentThread(iThread, SwitchReason.LOCK_WAIT, true)
             // Now it is possible to acquire the monitor, do it then.
-            monitorTracker.acquireMonitor(iThread, monitor)
+            require(monitorTracker.acquireMonitor(iThread, monitor))
         }
         // The monitor is acquired, finish.
         return false
@@ -499,7 +501,7 @@ abstract class ManagedStrategy(
      * @param iThread the number of the executed thread according to the [scenario][ExecutionScenario].
      * @param codeLocation the byte-code location identifier of this operation.
      * @param withTimeout `true` if is invoked with timeout, `false` otherwise.
-     * @return whether wait should be executed
+     * @return whether `Object.wait` should be executed
      */
     internal fun beforeWait(iThread: Int, codeLocation: Int, tracePoint: WaitTracePoint?, monitor: Any, withTimeout: Boolean): Boolean {
         if (!isTestThread(iThread)) return true
@@ -509,20 +511,23 @@ abstract class ManagedStrategy(
         monitorTracker.waitOnMonitor(iThread, monitor)
         // switch to another thread and wait till a notify event happens
         switchCurrentThread(iThread, SwitchReason.MONITOR_WAIT, true)
+        require(monitorTracker.acquireMonitor(iThread, monitor)) // acquire the lock again
         return false
     }
 
     /**
      * @param iThread the number of the executed thread according to the [scenario][ExecutionScenario].
      * @param codeLocation the byte-code location identifier of this operation.
+     * @return whether `Object.notify` should be executed
      */
-    internal fun afterNotify(iThread: Int, codeLocation: Int, tracePoint: NotifyTracePoint?, monitor: Any, notifyAll: Boolean) {
-        if (!isTestThread(iThread)) return
+    internal fun beforeNotify(iThread: Int, codeLocation: Int, tracePoint: NotifyTracePoint?, monitor: Any, notifyAll: Boolean): Boolean {
+        if (!isTestThread(iThread)) return true
         if (notifyAll)
             monitorTracker.notifyAll(monitor)
         else
             monitorTracker.notify(monitor)
         traceCollector?.passCodeLocation(tracePoint)
+        return false
     }
 
     /**

--- a/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategyTransformer.kt
+++ b/src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategyTransformer.kt
@@ -813,44 +813,54 @@ internal class ManagedStrategyTransformer(
      */
     private inner class WaitNotifyTransformer(methodName: String, mv: GeneratorAdapter) : ManagedStrategyMethodVisitor(methodName, mv) {
         override fun visitMethodInsn(opcode: Int, owner: String, name: String, desc: String, itf: Boolean) = adapter.run {
-            var afterWait: Label? = null
+            var skipWaitOrNotify = newLabel()
             val isWait = isWait(opcode, name, desc)
             val isNotify = isNotify(opcode, name, desc)
             if (isWait) {
-                afterWait = newLabel()
+                skipWaitOrNotify = newLabel()
                 val withTimeout = desc != "()V"
                 var lastArgument = 0
                 var firstArgument = 0
-                if (desc == "(J)V") {
-                    firstArgument = newLocal(Type.LONG_TYPE)
-                    storeLocal(firstArgument)
-                } else if (desc == "(JI)V") {
-                    lastArgument = newLocal(Type.INT_TYPE)
-                    storeLocal(lastArgument)
-                    firstArgument = newLocal(Type.LONG_TYPE)
-                    storeLocal(firstArgument)
+                when (desc) {
+                    "(J)V" -> {
+                        firstArgument = newLocal(Type.LONG_TYPE)
+                        storeLocal(firstArgument)
+                    }
+                    "(JI)V" -> {
+                        lastArgument = newLocal(Type.INT_TYPE)
+                        storeLocal(lastArgument)
+                        firstArgument = newLocal(Type.LONG_TYPE)
+                        storeLocal(firstArgument)
+                    }
                 }
-                dup()
+                dup() // copy monitor
                 invokeBeforeWait(withTimeout)
                 val beforeWait: Label = newLabel()
                 ifZCmp(GeneratorAdapter.GT, beforeWait)
-                pop()
-                goTo(afterWait)
+                pop() // pop monitor
+                goTo(skipWaitOrNotify)
                 visitLabel(beforeWait)
-                if (desc == "(J)V")
-                    loadLocal(firstArgument)
-                if (desc == "(JI)V") { // restore popped arguments
-                    loadLocal(firstArgument)
-                    loadLocal(lastArgument)
+                // restore popped arguments
+                when (desc) {
+                    "(J)V" -> loadLocal(firstArgument)
+                    "(JI)V" -> {
+                         loadLocal(firstArgument)
+                         loadLocal(lastArgument)
+                     }
                 }
             }
-            if (isNotify) dup()
-            visitMethodInsn(opcode, owner, name, desc, itf)
-            if (isWait) visitLabel(afterWait)
             if (isNotify) {
                 val notifyAll = name == "notifyAll"
-                invokeAfterNotify(notifyAll)
+                dup() // copy monitor
+                invokeBeforeNotify(notifyAll)
+                val beforeNotify = newLabel()
+                ifZCmp(GeneratorAdapter.GT, beforeNotify)
+                pop() // pop monitor
+                goTo(skipWaitOrNotify)
+                visitLabel(beforeNotify)
             }
+            visitMethodInsn(opcode, owner, name, desc, itf)
+            visitLabel(skipWaitOrNotify)
         }
 
         private fun isWait(opcode: Int, name: String, desc: String): Boolean {
@@ -874,8 +884,8 @@ internal class ManagedStrategyTransformer(
         }
 
         // STACK: monitor
-        private fun invokeAfterNotify(notifyAll: Boolean) {
-            invokeOnWaitOrNotify(AFTER_NOTIFY_METHOD, notifyAll, ::NotifyTracePoint, NOTIFY_TRACE_POINT_TYPE)
+        private fun invokeBeforeNotify(notifyAll: Boolean) {
+            invokeOnWaitOrNotify(BEFORE_NOTIFY_METHOD, notifyAll, ::NotifyTracePoint, NOTIFY_TRACE_POINT_TYPE)
         }
 
         // STACK: monitor
@@ -1307,7 +1317,7 @@ private val BEFORE_SHARED_VARIABLE_WRITE_METHOD = Method.getMethod(ManagedStrate
 private val BEFORE_LOCK_ACQUIRE_METHOD = Method.getMethod(ManagedStrategy::beforeLockAcquire.javaMethod)
 private val BEFORE_LOCK_RELEASE_METHOD = Method.getMethod(ManagedStrategy::beforeLockRelease.javaMethod)
 private val BEFORE_WAIT_METHOD = Method.getMethod(ManagedStrategy::beforeWait.javaMethod)
-private val AFTER_NOTIFY_METHOD = Method.getMethod(ManagedStrategy::afterNotify.javaMethod)
+private val BEFORE_NOTIFY_METHOD = Method.getMethod(ManagedStrategy::beforeNotify.javaMethod)
 private val BEFORE_PARK_METHOD = Method.getMethod(ManagedStrategy::beforePark.javaMethod)
 private val AFTER_UNPARK_METHOD = Method.getMethod(ManagedStrategy::afterUnpark.javaMethod)
 private val ENTER_IGNORED_SECTION_METHOD = Method.getMethod(ManagedStrategy::enterIgnoredSection.javaMethod)

--- a/src/jvm/test/org/jetbrains/kotlinx/lincheck/test/transformation/WaitNotifyLockTest.kt
+++ b/src/jvm/test/org/jetbrains/kotlinx/lincheck/test/transformation/WaitNotifyLockTest.kt
@@ -1,0 +1,73 @@
+/*
+ * Lincheck
+ *
+ * Copyright (C) 2019 - 2021 JetBrains s.r.o.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Lesser Public License for more details.
+ *
+ * You should have received a copy of the GNU General Lesser Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/lgpl-3.0.html>
+ */
+
+package org.jetbrains.kotlinx.lincheck.test.transformation
+
+import org.jetbrains.kotlinx.lincheck.LinChecker
+import org.jetbrains.kotlinx.lincheck.annotations.Operation
+import org.jetbrains.kotlinx.lincheck.strategy.managed.modelchecking.ModelCheckingCTest
+import org.jetbrains.kotlinx.lincheck.verifier.VerifierState
+import org.junit.Test
+
+@ModelCheckingCTest(iterations = 30)
+class WaitNotifyLockTest : VerifierState() {
+    private var counter = 0
+    private val lock = WaitNotifyLock()
+
+    override fun extractState() = counter
+
+    @Operation
+    fun getAndIncrement() = lock.withLock { counter++ }
+
+    @Test
+    fun test() {
+        LinChecker.check(this::class.java)
+    }
+}
+
+private class WaitNotifyLock {
+    private var owner: Thread? = null
+
+    fun lock() {
+        val thread = Thread.currentThread()
+        synchronized(this) {
+            while (owner != null) {
+                (this as Object).wait()
+            }
+            owner = thread
+        }
+    }
+
+    fun unlock() {
+        synchronized(this) {
+            owner = null
+            (this as Object).notify()
+        }
+    }
+}
+
+private inline fun <T> WaitNotifyLock.withLock(body: () -> T): T {
+    try {
+        lock()
+        return body()
+    } finally {
+        unlock()
+    }
+}

